data/src/main/java/com/microsoft/azure/kusto/data/auth/endpoints/KustoTrustedEndpoints.java

```diff
@@ -0,0 +1,144 @@
+            List<MatchRule> rules = new ArrayList<>();
+            value.AllowedKustoSuffixes.forEach(suffix -> rules.add(new MatchRule(suffix, false)));
+            value.AllowedKustoHostnames.forEach(hostname -> rules.add(new MatchRule(hostname, true)));
+            matchers.put(key, FastSuffixMatcher.Create(rules));
+        });
+
+        additionalMatcher = null;
+        overrideMatcher = null;
+    }
+
+    /**
+     * @param matcher Rules that determine if a hostname is a valid/trusted Kusto endpoint
+     *  (replaces existing rules). Null means "use the default policy".
+     */
+    public static void SetOverridePolicy(Predicate<String> matcher) {
+        overrideMatcher = matcher;
+    }
+
+    /**
+     * Throw an exception if the endpoint specified is not trusted.
+     *
+     * @param uri - Kusto endpoint
+     * @param loginEndpoint The login endpoint to check against.
+     * @throws KustoClientInvalidConnectionStringException - Endpoint is not a trusted Kusto endpoint
+     */
+    public static void ValidateTrustedEndpoint(String uri, String loginEndpoint) throws KustoClientInvalidConnectionStringException {
+        try {
+            ValidateTrustedEndpoint(new URI(uri), loginEndpoint);
+        } catch (URISyntaxException ex) {
+            throw new KustoClientInvalidConnectionStringException(ex);
+        }
+    }
+
+    /**
+     * Throw an exception if the endpoint specified is not trusted.
+     *
+     * @param uri - Kusto endpoint
+     * @throws KustoClientInvalidConnectionStringException - Endpoint is not a trusted Kusto endpoint
+     */
+    public static void ValidateTrustedEndpoint(String uri) throws KustoClientInvalidConnectionStringException {
+        try {
+            ValidateTrustedEndpoint(new URI(uri), CloudInfo.retrieveCloudInfoForCluster(uri).getLoginEndpoint());
+        } catch (URISyntaxException | DataServiceException ex) {
+            throw new KustoClientInvalidConnectionStringException(ex);
+        }
+    }
+
+    /**
+     * Is the endpoint uri trusted?
+     *
+     * @param uri The endpoint to inspect.
+     * @param loginEndpoint The login endpoint to check against.
+     * @throws KustoClientInvalidConnectionStringException - Endpoint is not a trusted Kusto endpoint
+     */
+    public static void ValidateTrustedEndpoint(URI uri, String loginEndpoint) throws KustoClientInvalidConnectionStringException {
+        Ensure.argIsNotNull(uri, "uri");
+
+        // Check that target hostname is trusted and can accept security token
+        ValidateHostnameIsTrusted(uri.getHost(), loginEndpoint);
+    }
+
+    /**
+     * Adds the rules that determine if a hostname is a valid/trusted Kusto endpoint
+     * (extends existing rules).
+     *
+     * @param rules   - A set of rules
+     * @param replace - If true nullifies the last added rules
+     */
+    public synchronized static void AddTrustedHosts(List<MatchRule> rules, boolean replace) {
```

Comment:
It's not synchronized when you use it

Reply:
what do you mean

Reply:
validateHostnameIsTrusted is not synchronized

Reply:
how about i change it to 
`
    public static void AddTrustedHosts(List<MatchRule> rules, boolean replace) {

        if (rules.isEmpty())
        {
            if (replace)
            {
                additionalMatcher = null;
            }
            return;
        }

        additionalMatcher = FastSuffixMatcher.Create(replace ? null : additionalMatcher, rules);
    }
`